blazar/plugins/oshosts/host_plugin.py

```diff
@@ -973,6 +973,39 @@ def poll_resource_failures(self):
                 recovered_hosts.extend([host for host in unreservable_hosts
                                         if host['id'] in active_hv_ids])
 
+            aggregates = self.nova.aggregates.list()
+            if aggregates:
+                freepool = next((agg for agg in aggregates if agg.name == "freepool"), None)
+                if not freepool:
+                    raise ValueError("Aggregate list does not contain a freepool!")
+                # Non-freepool aggregates are named after an associated reservation
+                res_ids = [agg.name for agg in aggregates if agg != freepool]
+                aggregate_reservations = {
+                    res["id"]: res for res in db_api.reservation_get_all_by_ids(res_ids)
+                }
+                for agg in aggregates:
+                    if agg == freepool:
+                        continue
+                    reservation = aggregate_reservations[agg.name]
+                    if reservation["status"] not in (status.reservation.ERROR, status.reservation.DELETED):
+                        # Any aggregates without an active reservation are the result of
+                        # a failed reservation teardown
+                        continue
+                    LOG.warning(
+                        f"Host aggregate for reservation {agg.name} failed to clean up."
+                    )
+                    for host in agg.hosts:
+                        try:
+                            # Remove the host from the reservation's aggregate
+                            agg.remove_host(host)
+                            # Return the host to the freepool
+                            freepool.add_host(host)
+                            recovered_hosts.append(host)
```

Comment:
I am wondering if it is possible for this to fail, and a node to be "lost" without an aggregate as a result.

Reply:
I think the reason this would fail is if the host is not found, or if the freepool already has a host. 
I am not sure how to enforce transaction operation here - maybe retry the operation?